### PR TITLE
DEVOPS-875-aggiornare-terraform-sh

### PR DESCRIPTION
**Pull request by fmlisco**
Here are the functions related to the script:
-**clean_environment**: removes the .terraform* folders and tfplan file in the current directory.
-**help_usage**: prints a help message with the usage of the script and the list of available actions.
-**init_terraform**: initializes the Terraform backend and modules.
-**list_env**: lists the names of the folders found in the env folder.
-**other_actions**: performs any Terraform action.
-**state_output_taint_actions**: performs the Terraform state, output, and taint actions.
-**tfsummary**: generates a summary of the Terraform plan using the tf-summarize library.
-**update_script**: checks if a new version of the script is available on GitHub. If the new version is available, it asks the user if they want to update the script and, if the user agrees, downloads and installs the new version.

In the following the commands you can use:
-**clean**: cleans up the environment by removing any temporary files or directories that were created during the execution of the script.
-**help** or ?: displays the usage information for the script, including a list of available actions and their descriptions.
-**init**: initializes the Terraform environment by running terraform init.
-**list**: lists all the available environments.
-**output, state, or taint**: run the terraform output, terraform state, or terraform taint commands
-**summ**: runs the tf-summarize command.
-**update_script**: updates the script to the latest version available on GitHub.  

**Please note** the `update_script` function is not complete since I need an uploaded version of the script for testing purposes

### CODE REVIEW

File: terraform.sh [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+    echo "curl not found, cannot update script"
+    exit 1
+  fi
+
+  current_version=$(sed -n '3p' "$0" | awk '{print $NF}')
+  repo_url="https://raw.githubusercontent.com/pagopa/eng-common-scripts/main/azure/terraform.sh"
+  response=$(curl -s -w "%{http_code}" "$repo_url")
+
+  if [ "${response: -3}" == "404" ]; then
+    echo "Script non trovato nel repository."
```
Comment (uolter) [inline terraform.sh]: English please :-)

File: terraform.sh [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+shift 2
+other=$@
+
+if [ -n "$env" ]; then
+  source "./env/$env/backend.ini"
+  if [ -z "$(command -v az)" ]; then
+    echo "az not found, cannot proceed"
+    exit 1
+  fi
+  az account set -s "${subscription}"
```
Comment (uolter) [inline terraform.sh]: Is this script intended to work with azure only?

**Review by pasqualedevita — APPROVED** [review]

File: terraform.sh [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+  exit 0
+fi
+
+# Parse arguments
+action=$1
+env=$2
+shift 2
+other=$@
+
+if [ -n "$env" ]; then
```
Comment (pasqualedevita) [inline terraform.sh]: this script is only for azure right? can we move it into azure dir?

**Review by pasqualedevita — CHANGES_REQUESTED** [review]

File: azure/migrate-resource.sh [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+  done
+  echo
+  echo "Syntax: setup.sh [-l|h]"
+  echo "  options:"
+  echo "  h     Print this Help."
+  echo "  l     List available environments."
+  echo
+}
+
+function removeAndImport() {
```
Comment (uolter) [inline azure/migrate-resource.sh:28]: Do we need this function? Is it doing this refactoring provided by terraform itself?
https://developer.hashicorp.com/terraform/language/modules/develop/refactoring

Comment (fmlisco) [reply azure/migrate-resource.sh:28]: They are in fact pretty similar but consider that the "moved blocks" functionality has some limitations; for example: "move" statements can be only within a single module package (https://nedinthecloud.com/2021/12/14/using-the-moved-block-in-terraform-1-1/)
Furthermore, consider it is not perfect (e.g. https://github.com/hashicorp/terraform/issues/32758 "Moved block failure after import"), so the reason for using this script is that you can customize the rename statement (aka removeandimport) via bash script or cli commands.

File: terraform.sh (hunk 3 above)
Comment (fmlisco) [reply terraform.sh]: done

File: terraform.sh (hunk 2 above)
Comment (fmlisco) [reply terraform.sh]: yep

File: terraform.sh (hunk 1 above)
Comment (fmlisco) [reply terraform.sh]: Right! Translated

File: terraform.sh (hunk 2 above)
Comment (fmlisco) [reply terraform.sh]: same as the original one. I kept the same functionalities extending it

**Review by pasqualedevita — APPROVED** [review]
LGTM

**Pull request merged**